apps/glusterfs/volume_entry_allocate.go

```diff
@@ -63,6 +66,226 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
+			return brickEntry, nil
+		}
+	}
+
+	return nil, ErrNotFound
+}
+
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
+	allocator Allocator,
+	oldBrickId string) (e error) {
+
+	var oldBrickEntry *BrickEntry
+	var oldDeviceEntry *DeviceEntry
+	var newDeviceEntry *DeviceEntry
+	var oldBrickNodeEntry *NodeEntry
+	var newBrickNodeEntry *NodeEntry
+
+	err := db.View(func(tx *bolt.Tx) error {
+		var err error
+		oldBrickEntry, err = NewBrickEntryFromId(tx, oldBrickId)
+		if err != nil {
+			return err
+		}
+
+		oldDeviceEntry, err = NewDeviceEntryFromId(tx, oldBrickEntry.Info.DeviceId)
+		if err != nil {
+			return err
+		}
+		oldBrickNodeEntry, err = NewNodeEntryFromId(tx, oldBrickEntry.Info.NodeId)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Determine the setlist by getting data from Gluster
+	vinfo, err := executor.VolumeInfo(oldBrickNodeEntry.ManageHostName(), v.Info.Name)
+	var slicestartindex int
+	var foundbrickset bool
+	var brick executors.Brick
+	setlist := make([]*BrickEntry, 0)
+	// BrickList in volume info is a slice of all bricks in volume
+	// We loop over the slice in steps of BricksInSet()
+	// If brick to be replaced is found in an iteration, other bricks in that slice form the setlist
+	for slicestartindex = 0; slicestartindex <= len(vinfo.Bricks.BrickList)-v.Durability.BricksInSet(); slicestartindex = slicestartindex + v.Durability.BricksInSet() {
+		setlist = make([]*BrickEntry, 0)
+		for _, brick = range vinfo.Bricks.BrickList[slicestartindex : slicestartindex+v.Durability.BricksInSet()] {
+			brickentry, err := v.getBrickEntryfromBrickName(db, brick.Name)
+			if err != nil {
+				logger.LogError("Unable to create brick entry using brick name:%v, error: %v", brick.Name, err)
+				return err
+			}
+			if brickentry.Id() == oldBrickId {
+				foundbrickset = true
+			} else {
+				setlist = append(setlist, brickentry)
+			}
+		}
+		if foundbrickset {
+			break
+		}
+	}
+	if !foundbrickset {
+		logger.LogError("Unable to find brick set for brick %v, db is possibly corrupt", oldBrickEntry.Id())
+		return ErrNotFound
+	}
+
+	//Create an Id for new brick
+	newBrickId := utils.GenUUID()
+
+	// Check the ring for devices to place the brick
+	deviceCh, done, errc := allocator.GetNodes(v.Info.Cluster, newBrickId)
+	defer func() {
+		close(done)
+	}()
+
+	for deviceId := range deviceCh {
+
+		// Get device entry
+		err = db.View(func(tx *bolt.Tx) error {
+			newDeviceEntry, err = NewDeviceEntryFromId(tx, deviceId)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		// Skip same device
+		if oldDeviceEntry.Info.Id == newDeviceEntry.Info.Id {
+			continue
+		}
+
+		// Do not allow a device from the same node to be
+		// in the set
+		deviceOk := true
+		for _, brickInSet := range setlist {
+			if brickInSet.Info.NodeId == newDeviceEntry.NodeId {
+				deviceOk = false
+			}
+		}
+
```

Comment:
Created #716 for zone logic.
Created #717 for moving the brick set determination logic to Upgrade() part.